// react-faq-component/build/index.js
'use strict';

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);
var PropTypes = _interopDefault(require('prop-types'));

function _classCallCheck(instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
}

function _defineProperties(target, props) {
  for (var i = 0; i < props.length; i++) {
    var descriptor = props[i];
    descriptor.enumerable = descriptor.enumerable || false;
    descriptor.configurable = true;
    if ("value" in descriptor) descriptor.writable = true;
    Object.defineProperty(target, descriptor.key, descriptor);
  }
}

function _createClass(Constructor, protoProps, staticProps) {
  if (protoProps) _defineProperties(Constructor.prototype, protoProps);
  if (staticProps) _defineProperties(Constructor, staticProps);
  return Constructor;
}

function _defineProperty(obj, key, value) {
  if (key in obj) {
    Object.defineProperty(obj, key, {
      value: value,
      enumerable: true,
      configurable: true,
      writable: true
    });
  } else {
    obj[key] = value;
  }

  return obj;
}

function _extends() {
  _extends = Object.assign || function (target) {
    for (var i = 1; i < arguments.length; i++) {
      var source = arguments[i];

      for (var key in source) {
        if (Object.prototype.hasOwnProperty.call(source, key)) {
          target[key] = source[key];
        }
      }
    }

    return target;
  };

  return _extends.apply(this, arguments);
}

function _inherits(subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function");
  }

  subClass.prototype = Object.create(superClass && superClass.prototype, {
    constructor: {
      value: subClass,
      writable: true,
      configurable: true
    }
  });
  if (superClass) _setPrototypeOf(subClass, superClass);
}

function _getPrototypeOf(o) {
  _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) {
    return o.__proto__ || Object.getPrototypeOf(o);
  };
  return _getPrototypeOf(o);
}

function _setPrototypeOf(o, p) {
  _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) {
    o.__proto__ = p;
    return o;
  };

  return _setPrototypeOf(o, p);
}

function _assertThisInitialized(self) {
  if (self === void 0) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }

  return self;
}

function _possibleConstructorReturn(self, call) {
  if (call && (typeof call === "object" || typeof call === "function")) {
    return call;
  }

  return _assertThisInitialized(self);
}

function styleInject(css, ref) {
  if (ref === void 0) ref = {};
  var insertAt = ref.insertAt;

  if (!css || typeof document === 'undefined') {
    return;
  }

  var head = document.head || document.getElementsByTagName('head')[0];
  var style = document.createElement('style');
  style.type = 'text/css';

  if (insertAt === 'top') {
    if (head.firstChild) {
      head.insertBefore(style, head.firstChild);
    } else {
      head.appendChild(style);
    }
  } else {
    head.appendChild(style);
  }

  if (style.styleSheet) {
    style.styleSheet.cssText = css;
  } else {
    style.appendChild(document.createTextNode(css));
  }
}

var css = ".styles_faq-row-wrapper__3Hsch {\n  background-color: var(--faq-bg-color, white); }\n  .styles_faq-row-wrapper__3Hsch h2 {\n    margin: 0;\n    color: var(--title-text-color, black); }\n  .styles_faq-row-wrapper__3Hsch .styles_faq-row__2Rd2Y {\n    display: flex;\n    justify-content: space-between;\n    padding: 5px 0;\n    border-bottom: 1px solid #ccc; }\n  .styles_faq-row-wrapper__3Hsch .styles_row-body__8wIE9 .styles_faq-row__2Rd2Y {\n    flex-direction: column; }\n    .styles_faq-row-wrapper__3Hsch .styles_row-body__8wIE9 .styles_faq-row__2Rd2Y .styles_row-title__2yqnt {\n      padding: 10px 0;\n      display: flex;\n      justify-content: space-between;\n      color: var(--row-title-color, black);\n      font-weight: 400;\n      font-size: var(--row-title-text-size, large);\n      letter-spacing: 0.25px;\n      cursor: pointer;\n      align-items: center; }\n      .styles_faq-row-wrapper__3Hsch .styles_row-body__8wIE9 .styles_faq-row__2Rd2Y .styles_row-title__2yqnt .styles_icon-wrapper__wjX56 {\n        max-width: 25px;\n        max-height: 25px;\n        margin: 0;\n        padding: 0;\n        color: var(--arrow-color, black); }\n        .styles_faq-row-wrapper__3Hsch .styles_row-body__8wIE9 .styles_faq-row__2Rd2Y .styles_row-title__2yqnt .styles_icon-wrapper__wjX56 .styles_arrow-image__qLm48,\n        .styles_faq-row-wrapper__3Hsch .styles_row-body__8wIE9 .styles_faq-row__2Rd2Y .styles_row-title__2yqnt .styles_icon-wrapper__wjX56 svg {\n          width: 100%;\n          height: 100%; }\n        .styles_faq-row-wrapper__3Hsch .styles_row-body__8wIE9 .styles_faq-row__2Rd2Y .styles_row-title__2yqnt .styles_icon-wrapper__wjX56 svg {\n          fill: var(--arrow-color, black); }\n      .styles_faq-row-wrapper__3Hsch .styles_row-body__8wIE9 .styles_faq-row__2Rd2Y .styles_row-title__2yqnt.styles_closed__3LKOc + .styles_row-content__TVd0Y {\n        visibility: hidden;\n        z-index: -1; }\n        .styles_faq-row-wrapper__3Hsch .styles_row-body__8wIE9 .styles_faq-row__2Rd2Y .styles_row-title__2yqnt.styles_closed__3LKOc + .styles_row-content__TVd0Y.styles_animate__3cloa {\n          opacity: 0;\n          transition: height 0.12s, opacity 0.12s; }\n        .styles_faq-row-wrapper__3Hsch .styles_row-body__8wIE9 .styles_faq-row__2Rd2Y .styles_row-title__2yqnt.styles_closed__3LKOc + .styles_row-content__TVd0Y.styles_static__1EHAK {\n          display: none; }\n        .styles_faq-row-wrapper__3Hsch .styles_row-body__8wIE9 .styles_faq-row__2Rd2Y .styles_row-title__2yqnt.styles_closed__3LKOc + .styles_row-content__TVd0Y .styles_row-content-text__2B1BC {\n          margin: 0 0; }\n      .styles_faq-row-wrapper__3Hsch .styles_row-body__8wIE9 .styles_faq-row__2Rd2Y .styles_row-title__2yqnt.styles_closed__3LKOc .styles_icon-wrapper__wjX56 {\n        transform: rotate(0deg);\n        transition: transform 0.12s; }\n      .styles_faq-row-wrapper__3Hsch .styles_row-body__8wIE9 .styles_faq-row__2Rd2Y .styles_row-title__2yqnt.styles_expanded__2KFpr + .styles_row-content__TVd0Y {\n        visibility: visible;\n        z-index: 0; }\n        .styles_faq-row-wrapper__3Hsch .styles_row-body__8wIE9 .styles_faq-row__2Rd2Y .styles_row-title__2yqnt.styles_expanded__2KFpr + .styles_row-content__TVd0Y.styles_animate__3cloa {\n          opacity: 1;\n          transition: height 0.12s, opacity 0.12s; }\n        .styles_faq-row-wrapper__3Hsch .styles_row-body__8wIE9 .styles_faq-row__2Rd2Y .styles_row-title__2yqnt.styles_expanded__2KFpr + .styles_row-content__TVd0Y.styles_static__1EHAK {\n          display: block; }\n        .styles_faq-row-wrapper__3Hsch .styles_row-body__8wIE9 .styles_faq-row__2Rd2Y .styles_row-title__2yqnt.styles_expanded__2KFpr + .styles_row-content__TVd0Y .styles_row-content-text__2B1BC {\n          margin: var(--row-content-padding-top, 0) var(--row-content-padding-right, 0) var(--row-content-padding-bottom, 0) var(--row-content-padding-left, 0); }\n      .styles_faq-row-wrapper__3Hsch .styles_row-body__8wIE9 .styles_faq-row__2Rd2Y .styles_row-title__2yqnt.styles_expanded__2KFpr .styles_icon-wrapper__wjX56 {\n        transform: rotate(180deg);\n        transition: transform 0.12s; }\n    .styles_faq-row-wrapper__3Hsch .styles_row-body__8wIE9 .styles_faq-row__2Rd2Y .styles_row-content-text__2B1BC {\n      color: var(--row-content-color, black);\n      font-size: var(--row-content-text-size, medium); }\n";
var style = {"faq-row-wrapper":"styles_faq-row-wrapper__3Hsch","faq-row":"styles_faq-row__2Rd2Y","row-body":"styles_row-body__8wIE9","row-title":"styles_row-title__2yqnt","icon-wrapper":"styles_icon-wrapper__wjX56","arrow-image":"styles_arrow-image__qLm48","closed":"styles_closed__3LKOc","row-content":"styles_row-content__TVd0Y","animate":"styles_animate__3cloa","static":"styles_static__1EHAK","row-content-text":"styles_row-content-text__2B1BC","expanded":"styles_expanded__2KFpr"};
styleInject(css);

var arrow_down = "<svg xmlns=\"http://www.w3.org/2000/svg\" viewBox=\"0 0 24 24\" fill=\"black\" width=\"36px\" height=\"36px\"><path d=\"M7.41 8.59L12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z\"/><path d=\"M0 0h24v24H0V0z\" fill=\"none\"/></svg>";

var rowItem =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(rowItem, _PureComponent);

  function rowItem() {
    var _getPrototypeOf2;

    var _this;

    _classCallCheck(this, rowItem);

    for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
      args[_key] = arguments[_key];
    }

    _this = _possibleConstructorReturn(this, (_getPrototypeOf2 = _getPrototypeOf(rowItem)).call.apply(_getPrototypeOf2, [this].concat(args)));

    _defineProperty(_assertThisInitialized(_this), "state", {
      isExpanded: false,
      ref: React__default.createRef(),
      height: 0
    });

    _defineProperty(_assertThisInitialized(_this), "expand", function () {
      var _this$props$config = _this.props.config;
      _this$props$config = _this$props$config === void 0 ? {} : _this$props$config;
      var _this$props$config$an = _this$props$config.animate,
          animate = _this$props$config$an === void 0 ? true : _this$props$config$an;

      _this.setState(function (prevState) {
        return {
          isExpanded: !prevState.isExpanded
        };
      }, animate ? _this.setHeight : undefined);
    });

    _defineProperty(_assertThisInitialized(_this), "setHeight", function () {
      setTimeout(function () {
        var _this$state = _this.state,
            ref = _this$state.ref,
            isExpanded = _this$state.isExpanded;
        var height = "calc(" + ref.current.scrollHeight + "px" + " + " + (_this.props.rowContentPaddingTop ? _this.props.rowContentPaddingTop : "0px") + " + " + (_this.props.rowContentPaddingBottom ? _this.props.rowContentPaddingBottom : "0px") + ")";

        _this.setState({
          height: isExpanded ? height : 0
        });
      }, 1);
    });

    return _this;
  }

  _createClass(rowItem, [{
    key: "render",
    value: function render() {
      var _this$props = this.props,
          _this$props$data = _this$props.data,
          title = _this$props$data.title,
          content = _this$props$data.content,
          _this$props$config2 = _this$props.config;
      _this$props$config2 = _this$props$config2 === void 0 ? {} : _this$props$config2;
      var _this$props$config2$a = _this$props$config2.animate,
          animate = _this$props$config2$a === void 0 ? true : _this$props$config2$a,
          arrowIcon = _this$props$config2.arrowIcon;
      var _this$state2 = this.state,
          isExpanded = _this$state2.isExpanded,
          ref = _this$state2.ref,
          height = _this$state2.height;
      var attrs = {
        onClick: this.expand,
        "role": "button",
        "aria-expanded": isExpanded,
        "aria-controls": "react-faq-rowcontent-".concat(this.props.rowid)
      };
      var contentAttrs = {
        "role": "region",
        "id": "react-faq-rowcontent-".concat(this.props.rowid),
        "aria-expanded": isExpanded,
        "aria-hidden": !isExpanded
      };

      if (animate) {
        contentAttrs['style'] = {
          height: height
        };
      }

      var className = ['row-title', isExpanded ? 'expanded' : 'closed', style['row-title'], style[isExpanded ? 'expanded' : 'closed']].join(' ');
      var icon = arrowIcon || React__default.createElement("div", {
        dangerouslySetInnerHTML: {
          __html: arrow_down
        },
        className: "arrow-image ".concat(style["arrow-image"]),
        alt: "Expand arrow"
      });
      var contentClasses = [style["row-content"], "row-content", animate ? style["animate"] : style["static"]].join(' ');
      var contentTextClasses = [style["row-content-text"], "row-content-text"].join(' ');
      return React__default.createElement("section", {
        className: "faq-row ".concat(style["faq-row"])
      }, React__default.createElement("div", _extends({
        className: className
      }, attrs, {
        role: "button"
      }), React__default.createElement("div", null, title), React__default.createElement("span", {
        className: "icon-wrapper ".concat(style["icon-wrapper"]),
        "aria-hidden": "true"
      }, icon)), React__default.createElement("div", _extends({
        className: contentClasses
      }, contentAttrs), React__default.createElement("div", {
        className: contentTextClasses,
        ref: ref,
        dangerouslySetInnerHTML: {
          __html: content
        }
      })));
    }
  }]);

  return rowItem;
}(React.PureComponent);

var FaqComponent =
/*#__PURE__*/
function (_PureComponent) {
  _inherits(FaqComponent, _PureComponent);

  function FaqComponent() {
    _classCallCheck(this, FaqComponent);

    return _possibleConstructorReturn(this, _getPrototypeOf(FaqComponent).apply(this, arguments));
  }

  _createClass(FaqComponent, [{
    key: "render",
    value: function render() {
      var _this = this;

      var _ref = this.props.data || {},
          title = _ref.title,
          _ref$rows = _ref.rows,
          rows = _ref$rows === void 0 ? [] : _ref$rows;

      var _this$props = this.props,
          _this$props$styles = _this$props.styles,
          styles = _this$props$styles === void 0 ? {} : _this$props$styles,
          _this$props$config = _this$props.config;
      _this$props$config = _this$props$config === void 0 ? {} : _this$props$config;
      var animate = _this$props$config.animate;
      var styleConfig = {
        "--faq-bg-color": styles.bgColor,
        "--title-text-color": styles.titleTextColor,
        "--row-title-color": styles.rowTitleColor,
        "--row-title-text-size": styles.rowTitleTextSize,
        "--row-content-color": styles.rowContentColor,
        "--row-content-text-size": styles.rowContentTextSize,
        "--row-content-padding-top": styles.rowContentPaddingTop,
        "--row-content-padding-bottom": styles.rowContentPaddingBottom,
        "--row-content-padding-right": styles.rowContentPaddingRight,
        "--row-content-padding-left": styles.rowContentPaddingLeft,
        "--arrow-color": styles.arrowColor
      };
      var wrapperClass = "faq-row-wrapper ".concat(style["faq-row-wrapper"]);
      var titleClass = "faq-title ".concat(style["faq-row"]);
      var rowlistClass = "faq-body ".concat(style["row-body"]);
      return React__default.createElement("div", {
        className: wrapperClass,
        style: styleConfig
      }, title ? React__default.createElement("section", {
        className: titleClass
      }, React__default.createElement("h2", null, title)) : null, rows.length ? React__default.createElement("section", {
        className: rowlistClass
      }, rows.map(function (row, i) {
        return React__default.createElement(rowItem, {
          data: row,
          key: i,
          rowid: i + 1,
          config: _this.props.config,
          rowContentPaddingTop: styles.rowContentPaddingTop,
          rowContentPaddingBottom: styles.rowContentPaddingBottom
        });
      })) : null);
    }
  }]);

  return FaqComponent;
}(React.PureComponent);

_defineProperty(FaqComponent, "propTypes", {
  data: PropTypes.object
});

module.exports = FaqComponent;
